homeassistant/components/esphomelib/config_flow.py

```diff
@@ -0,0 +1,129 @@
+"""Config flow to configure esphomelib component."""
+import asyncio
+from collections import OrderedDict
+from typing import Optional
+
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.helpers import ConfigType
+
+
+@config_entries.HANDLERS.register('esphomelib')
+class EsphomelibFlowHandler(config_entries.ConfigFlow):
+    """Handle a esphomelib config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    def __init__(self):
+        """Initialize flow."""
+        self._host = None  # type: Optional[str]
+        self._port = None  # type: Optional[int]
+        self._password = None  # type: Optional[str]
+        self._name = None  # type: Optional[str]
+
+    async def async_step_user(self, user_input: Optional[ConfigType] = None,
+                              error: Optional[str] = None):
+        """Handle a flow initialized by the user."""
+        if user_input is not None:
+            self._host = user_input['host']
+            self._port = user_input['port']
+            error, device_info = await fetch_device_info(self._host,
+                                                         self._port)
+            if error is not None:
+                return await self.async_step_user(error=error)
+            self._name = device_info.name
+
+            # Only show authentication step if device uses password
+            if device_info.uses_password:
+                return await self.async_step_authenticate()
+
+            return self._async_get_entry()
+
+        fields = OrderedDict()
+        fields[vol.Required('host', default=self._host or vol.UNDEFINED)] = str
+        fields[vol.Optional('port', default=self._port or 6053)] = int
+
+        errors = {}
+        if error is not None:
+            errors['base'] = error
+
+        return self.async_show_form(
+            step_id='user',
+            data_schema=vol.Schema(fields),
+            errors=errors
+        )
+
+    def _async_get_entry(self):
+        return self.async_create_entry(
+            title=self._name,
+            data={
+                'host': self._host,
+                'port': self._port,
+                # The API uses protobuf, so empty string denotes absence
+                'password': self._password or '',
```

Comment:
It's not possible to avoid storing the password?

Reply:
Sure but there's not really a need for such logic IMHO. Otherwise this logic would have to sit in `async_setup_entry` like this:

```yaml
password = entry.data.get(CONF_PASSWORD, '')
```

but IMO that kind of defeats the purpose of config entries (having a well-defined internal data schema that does not need any validation).

Reply:
I mean, would it be possible to avoid storing the password at all, and just use what the user enters in the form to do the authentication at that time, and store something else (token) that represents the user? I'm thinking about how hue or tradfri integrations work. I'm guessing the auth schema of esplib doesn't allow that?

Reply:
I think what Martin means is that the username and password are exchanged for a token that is then used to make requests. That's how most web services, like oauth2, work.

Reply:
No the API doesn't allow that in this case and I can't really do anything against that.

There's a few reasons for that:

 - The ESPs have basically no storage, so they can't store any tokens.
 - Additionally, the password will be stored in a YAML file somewhere anyway. And if the user is using the esphomeyaml Hass.io add-on that path will be `<HA_CONFIG_DIR>/esphomeyaml/<FILENAME>`. So if someone has access to HA's storage dir then they will have the password anyway.